### PR TITLE
Deprecates the module cache helper function

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for pr #40316.

### Summary of Changes
It adds a new `loadFromCache` function in the `AbstractModuleDispatcher` as replacement for the static `ModuleHelper::moduleCache` function

### Testing Instructions
- Create some articles
- Create an instance of the module Articles Category
- Open the front end

### Actual result BEFORE applying this Pull Request
It shows some articles in the module.

### Expected result AFTER applying this Pull Request
It shows some articles in the module.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/103
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: libraries/src/Dispatcher/AbstractModuleDispatcher.php [hunk 1]
```diff
@@ -10,9 +10,13 @@
 namespace Joomla\CMS\Dispatcher;
 
 use Joomla\CMS\Application\CMSApplicationInterface;
+use Joomla\CMS\Cache\CacheControllerFactoryAwareInterface;
+use Joomla\CMS\Cache\CacheControllerFactoryAwareTrait;
+use Joomla\CMS\Filter\InputFilter;
 use Joomla\CMS\Helper\ModuleHelper;
 use Joomla\Input\Input;
 use Joomla\Registry\Registry;
+use stdClass;
```
Comment (HLeithner) [inline libraries/src/Dispatcher/AbstractModuleDispatcher.php]:
```suggestion
```

**Comment by joomdonation** [comment]
I don't know how this replacement could be used to cache the output of the whole module in `ModuleRenderer` here https://github.com/joomla/joomla-cms/blob/4.3-dev/libraries/src/Document/Renderer/Html/ModuleRenderer.php#L83 ?

**Comment by laoneo** [comment]
Why not, what is different to what we have now in the abstract class?

**Comment by joomdonation** [comment]
There is no different behavior in the code. However, from what I see, the `loadFromCache` method from abstract can only be used in the same way with cache callback controller, that is to cache the return values of a method. This works for modules implement it own cache like **mod_articles_categories**, **mod_articles_category**, **mod_related_items**.....

There are other modules which we need to cache it's whole output. That's the logic implemented in `ModuleRenderer` which I mentioned above. How would you handle that using the new code? Maybe I do not see something ?

Maybe a simple question: How you handle cache for `mod_articles_latest` use the this replacement?

**Comment by laoneo** [comment]
The cache in ModuleRenderer caches the whole output of the module while this one here calls the module and caches only the data. So on some point this caching should be replaced with the one in ModuleRenderer. Means they are completely different.

**Comment by laoneo** [comment]
Done

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on e8db14ab230bcb73aa5f077cf9f3d0a1c6ad72e9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40325">issues.joomla.org/tracker/joomla-cms/40325</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on f301c61d76f08a3eff51d4afdcfde8833f7e7c0a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40325">issues.joomla.org/tracker/joomla-cms/40325</a>.</sub>

File: modules/mod_articles_category/src/Dispatcher/Dispatcher.php [hunk 2]
```diff
@@ -68,14 +68,14 @@ protected function getLayoutData(): array
...
-        $cacheParams->class        = $this->getHelperFactory()->getHelper('ArticlesCategoryHelper');
-        $cacheParams->method       = 'getArticles';
-        $cacheParams->methodparams = [$params, $data['app']];
-        $cacheParams->modeparams   = md5(serialize([$idBase, $this->module->module, $this->module->id]));
+        $cacheParams = new Registry();
+        $cacheParams->set('cachemode', 'id');
+        $cacheParams->set('class', $this->getHelperFactory()->getHelper('ArticlesCategoryHelper'));
+        $cacheParams->set('method', 'getArticles');
+        $cacheParams->set('methodparams', [$params, $data['app']]);
+        $cacheParams->set('modeparams', md5(serialize([$idBase, $this->module->module, $this->module->id])));
```
Comment (HLeithner) [inline modules/mod_articles_category/src/Dispatcher/Dispatcher.php:76]: Does this really make sense to use our expensive Registry for cachhing here? wouldn't a simply array fit too?

**Comment by Fedik** [comment]
I am sorry to say, 
We cannot deprecate `ModuleHelper::moduleCache` because it actively used for module caching, when global cache is enabled.
https://github.com/joomla/joomla-cms/blob/1e7527b60ec42032abe7bfb9c000bacf38f65845/libraries/src/Document/Renderer/Html/ModuleRenderer.php#L93

Additionally `loadFromCache` introduce a large code duplication.
It does not looks right to me.

**Pull request closed without merging**